src/components/Auth/UserFormDialog/controls/MultiFactorAuth.tsx

```diff
@@ -0,0 +1,120 @@
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+import { ListDivider } from '@rmwc/list';
+import { Typography } from '@rmwc/typography';
+import React, { useEffect } from 'react';
+import { useFieldArray } from 'react-hook-form';
+import { FormContextValues } from 'react-hook-form/dist/contextTypes';
+
+import { CheckboxField, Field } from '../../../common/Field';
+import { AddAuthUserPayload, AuthFormUser } from '../../types';
+import styles from './controls.module.scss';
+import PhoneControl from './PhoneControl';
+
+// Consistent with the back-end validation. We want to be as loose as possible in
+// the emulator to avoid false negatives.
+const PHONE_REGEX = /^\+/;
+
+export type MultiFactorProps = {
+  user?: AddAuthUserPayload;
+};
+
+export const MultiFactor: React.FC<
+  MultiFactorProps & FormContextValues<AuthFormUser>
+> = (form) => {
+  const {
+    control,
+    watch,
+    setError,
+    clearError,
+    formState,
+    errors,
+    user,
+    register,
+    setValue,
+  } = form;
+
+  // https://react-hook-form.com/v5/api#useFieldArray
+  const { fields, append } = useFieldArray({
+    control,
+    name: 'mfaInfo',
+  });
+
+  const emailVerified = watch('emailVerified');
+  const mfaEnabled = watch('mfaEnabled');
+  const mfaInfo = watch('mfaInfo');
+
+  function getErrorText() {
+    if (errors.phoneNumber) {
+      if (errors.phoneNumber.type === 'pattern') {
+        return 'Phone number must be in international format and start with a "+"';
+      }
+    }
+  }
+
+  function addNewMfaNumber() {
+    append({
+      enrolledAt: new Date().toISOString(),
+      phoneInfo: '',
+      mfaEnrollmentId: 'PLACEHOLDER' + Math.random() * 200,
+    });
+  }
+
+  return (
+    <div className={styles.signInWrapper}>
+      <ListDivider tag="div" />
+      <div className={styles.signInHeader}>
+        <Typography use="body1" theme="textPrimaryOnBackground">
+          Multi-factor Authentication
+        </Typography>
+      </div>
+      <CheckboxField
+        name="mfaEnabled"
+        label="Enabled"
+        defaultChecked={false}
+        disabled={!emailVerified || emailVerified.length === 0}
+        inputRef={register()}
+      />
+
+      <Typography use="body2" tag="div">
+        SMS settings
+      </Typography>
+      {mfaEnabled &&
+        mfaEnabled.length > 0 &&
+        fields.map((item, index) => {
+          return (
+            <Field
+              key={item.id}
+              name={`mfaInfo.${index}.phoneInfo`}
+              label="Phone"
+              placeholder="Enter phone number"
+              type="tel"
+              error={getErrorText()}
+              inputRef={register({ pattern: PHONE_REGEX })}
```

Comment:
@yuchenshi I'm stuck on some really weird behavior by the `react-hook-form` library. All fields besides `phoneInfo` for a given `MfaEnrollment` in the `mfaInfo` array get removed, leaving only `phoneInfo`. Any guess at what could be going on? I've been trying to debug this all day.

Reply:
Maybe `useFieldArray` doesn't handle nested properties that are not inputs? I think we'd better try some workarounds like using a different array and converting on `save` just like how you've handled the checkbox conversion.

Reply:
Thanks, I've updated it to do that and it's working now. Frustrating that the form library mysteriously removed those fields.